### PR TITLE
feat: support login by following wechat official account

### PR DESCRIPTION
**Pull request by forestmgy**
Fix: https://github.com/casdoor/casdoor/issues/1251

demo site here : http://124.221.198.246:7001/login

### CODE REVIEW

File: idp/wechat.go [hunk 1]
```diff
@@ -191,3 +195,52 @@ func (idp *WeChatIdProvider) GetUserInfo(token *oauth2.Token) (*UserInfo, error)
 	}
 	return &userInfo, nil
 }
+
+func GetWechatOfficialAccountAccessToken(clientId string, clientSecret string) string {
+	request, err := http.NewRequest("GET", "https://api.weixin.qq.com/cgi-bin/token?grant_type=client_credential&appid="+clientId+"&secret="+clientSecret, nil)
```
Comment (hsluoyz) [inline idp/wechat.go:200]: fmt.Sprintf

File: idp/wechat.go [hunk 2]
```diff
@@ -191,3 +195,52 @@ func (idp *WeChatIdProvider) GetUserInfo(token *oauth2.Token) (*UserInfo, error)
...
+	}
+	return data.AccessToken
+}
+
+func GetWechatOfficialAccountQRCode(clientId string, clientSecret string) string {
+	accessToken := GetWechatOfficialAccountAccessToken(clientId, clientSecret)
+	client := new(http.Client)
+	params := "{\"expire_seconds\": 604800, \"action_name\": \"QR_STR_SCENE\", \"action_info\": {\"scene\": {\"scene_str\": \"test\"}}}"
+	bodyData := bytes.NewReader([]byte(params))
+	requeset, err := http.NewRequest("POST", "https://api.weixin.qq.com/cgi-bin/qrcode/create?access_token="+accessToken, bodyData)
```
Comment (hsluoyz) [inline idp/wechat.go:223]: fmt.Sprintf

File: idp/wechat.go [hunk 3]
```diff
@@ -191,3 +195,52 @@ func (idp *WeChatIdProvider) GetUserInfo(token *oauth2.Token) (*UserInfo, error)
...
+	if err != nil {
+		log.Println(err.Error())
+	}
+	return data.AccessToken
+}
+
+func GetWechatOfficialAccountQRCode(clientId string, clientSecret string) string {
+	accessToken := GetWechatOfficialAccountAccessToken(clientId, clientSecret)
+	client := new(http.Client)
+	params := "{\"expire_seconds\": 604800, \"action_name\": \"QR_STR_SCENE\", \"action_info\": {\"scene\": {\"scene_str\": \"test\"}}}"
```
Comment (hsluoyz) [inline idp/wechat.go:221]: Static string?

File: object/application.go [hunk 4]
```diff
@@ -119,9 +120,9 @@ func getProviderMap(owner string) map[string]*Provider {
 	providers := GetProviders(owner)
 	m := map[string]*Provider{}
 	for _, provider := range providers {
-		//if provider.Category != "OAuth" {
-		//	continue
-		//}
+		if provider.Type == "WeChat" {
+			provider.WeChatQRCode = idp.GetWechatOfficialAccountQRCode(provider.ClientId2, provider.ClientSecret2)
```
Comment (hsluoyz) [inline object/application.go:124]: Request this every time?

File: object/provider.go [hunk 5]
```diff
@@ -37,19 +37,21 @@ type Provider struct {
...
-	Title      string `xorm:"varchar(100)" json:"title"`
-	Content    string `xorm:"varchar(1000)" json:"content"`
-	Receiver   string `xorm:"varchar(100)" json:"receiver"`
+	Host                            string `xorm:"varchar(100)" json:"host"`
+	Port                            int    `json:"port"`
+	DisableSsl                      bool   `json:"disableSsl"`
+	Title                           string `xorm:"varchar(100)" json:"title"`
+	Content                         string `xorm:"varchar(1000)" json:"content"`
+	Receiver                        string `xorm:"varchar(100)" json:"receiver"`
+	MustFollowWechatOfficialAccount bool   `json:"mustFollowWechatOfficialAccount"`
```
Comment (hsluoyz) [inline object/provider.go:54]: Re-use DisableSsl

File: object/provider.go [hunk 6]
```diff
@@ -221,6 +223,7 @@ func UpdateProvider(id string, provider *Provider) bool {
 	if provider.ClientSecret2 == "***" {
 		session = session.Omit("client_secret2")
 	}
+
```
Comment (hsluoyz) [inline object/provider.go:226]: Dont change this

File: routers/router.go [hunk 7]
```diff
@@ -54,6 +54,8 @@ func initAPI() {
 	beego.Router("/api/get-saml-login", &controllers.ApiController{}, "GET:GetSamlLogin")
 	beego.Router("/api/acs", &controllers.ApiController{}, "POST:HandleSamlLogin")
 	beego.Router("/api/saml/metadata", &controllers.ApiController{}, "GET:GetSamlMeta")
+	beego.Router("/api/wechat", &controllers.ApiController{}, "POST:GetOfficialAccountEvent")
```
Comment (hsluoyz) [inline routers/router.go:57]: Why two new APIs?

File: web/src/auth/ProviderButton.js [hunk 8]
```diff
@@ -116,11 +118,33 @@ function getSamlUrl(provider, location) {
...
-      return (
-        <a key={provider.displayName} href={Provider.getAuthUrl(application, provider, "signup")}>
-          <img width={width} height={width} src={getProviderLogoURL(provider)} alt={provider.displayName} style={{margin: margin}} />
-        </a>
-      );
+      if (provider.type === "WeChat" && provider.clientId2 !== "" && provider.clientSecret2 !== "" && provider.weChatQRCode !== "" && provider.mustFollowWechatOfficialAccount === true) {
+        const info = async() => {
+          const t1 = setInterval(await getEvent, 3000, application, provider);
+          {Modal.info({
+            title: "Please use your mobile phone scan this QR code and then follow the Official Account",
```
Comment (hsluoyz) [inline web/src/auth/ProviderButton.js:125]: i18n, "Please use WeChat and scan the QR code to sign in"

File: web/src/locales/de/data.json [hunk 9]
```diff
@@ -491,6 +491,7 @@
     "Metadata - Tooltip": "Metadaten - Tooltip",
     "Method": "Methode",
     "Method - Tooltip": "Login behaviors, QR code or silent authorization",
+    "Must follow wechat official account - Tooltip": "Must follow wechat official account - Tooltip",
```
Comment (hsluoyz) [inline web/src/locales/de/data.json:494]: "Enable QR code"

File: controllers/auth.go [hunk 10]
```diff
@@ -531,3 +536,39 @@ func (c *ApiController) HandleSamlLogin() {
...
+	var data struct {
+		MsgType string `xml:"MsgType"`
+		Event   string `xml:"Event"`
+	}
+	err = xml.Unmarshal(respBytes, &data)
+	if err != nil {
+		fmt.Printf("error: %v", err)
+		return
+	}
+	wechatScanType["event"] = data.Event
```
Comment (hsluoyz) [inline controllers/auth.go:558]: 1. Only one key, why use map?

2. concurrent issue?

File: controllers/auth.go [hunk 11]
```diff
@@ -531,3 +536,39 @@ func (c *ApiController) HandleSamlLogin() {
...
 }
+
+// GetOfficialAccountEvent ...
+// @Tag GetOfficialAccountEvent API
+// @Title GetOfficialAccountEvent
+// @router /api/wechat [POST]
+func (c *ApiController) GetOfficialAccountEvent() {
+	respBytes, err := ioutil.ReadAll(c.Ctx.Request.Body)
+	if err != nil {
+		log.Println(err.Error())
```
Comment (hsluoyz) [inline controllers/auth.go:547]: Do we handle error this way?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: idp/wechat.go (hunk 3 above)
Comment (forestmgy) [reply idp/wechat.go:221]: I use the example in Wechat docs
image

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**